### PR TITLE
[BENCHMARKS] Replace XE-TLA with SYCL-TLA generic gemm + OneDNN

### LINKED ISSUE #6840: Stream-K GEMM benchmark — replace xetla provider with sycl-tla
# Stream-K GEMM benchmark — replace xetla provider with sycl-tla

**Parent**: #6724 — Switch XETLA benchmarks to SYCL_TLA analogues
**Complexity**: Low (no new C++ code needed; stream-k is not available in SYCL-TLA and cannot be added)
**Dependencies**: None

## Current state

`benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py` uses xetla as the only non-Triton provider:

```python
from triton_kernels_benchmark import xetla_kernel
```

Providers: `triton`, `xetla`

Benchmark shapes (1 total):
- `3072 x 4096 x 3072` (M x K x N)

The xetla stream-k provider (lines 274-285):
- Calls `xetla_kernel.gemm_streamk_shape_3072_4096_3072`
- Has `assert_close` commented out
- Requires extra `acc` and `cnt` tensors

Note: The main `gemm_benchmark.py` has a FIXME comment (line 293) indicating this shape is temporarily kept in the standard GEMM benchmark until stream-k gets better performance.

## SYCL-TLA documentation cross-reference

**Source**: [intel/sycl-tla docs](https://github.com/intel/sycl-tla/tree/main/media/docs/cpp)

### Stream-K support in SYCL-TLA / CUTLASS
- `gemm_api.md` documents four device-level GEMM operators: standard, batched-array, batched-strided, and split-k parallel. **Stream-K is NOT listed.**
- `efficient_gemm.md` does not mention stream-k at all — only standard tiling, split-k, and sliced-k strategies.
- `gemm_api_3x.md` (CUTLASS 3.x API) describes kernel scheduling policies (`KernelTmaWarpSpecialized*`, `KernelCpAsyncWarpSpecialized*`) but does not include a stream-k schedule.
- `grouped_scheduler.md` describes persistent kernel scheduling for grouped GEMM but uses round-robin tile assignment, not stream-k work-stealing.
- **Conclusion**: Stream-K is not available in SYCL-TLA. The CUTLASS framework that SYCL-TLA is based on does not implement stream-k scheduling. NVIDIA's upstream CUTLASS added stream-k later (in separate examples), but the Intel SYCL fork does not include it.

### Implications for this benchmark
Since neither split-k nor stream-k exist in SYCL-TLA, the replacement for the xetla stream-k provider must be the generic SYCL-TLA GEMM. This is still a meaningful comparison: stream-k's value proposition is specifically that it handles shapes with poor tile quantization better than standard GEMM. Comparing Triton stream-k vs SYCL-TLA standard GEMM directly measures this.

## Key decision needed

**SYCL-TLA does NOT have a Stream-K GEMM kernel, and unlike split-k (which has `GemmSplitKParallel` in the framework), there is no underlying CUTLASS support for stream-k at all.** A dedicated SYCL-TLA stream-k kernel is not feasible without writing a custom scheduling algorithm from scratch.

### Option A: Use generic SYCL-TLA GEMM + onednn as dual baselines (recommended)
- Compare Triton stream-k against `sycl_tla_kernel.gemm` (standard tiled GEMM) and `torch.matmul` (onednn)
- Pro: Zero new C++ code. Directly measures stream-k's value: "does Triton stream-k beat standard GEMM on shapes with poor tile quantization?"
- Pro: The benchmark shape `3072x4096x3072` is specifically chosen because it doesn't divide evenly by large tile sizes (e.g., 256x256) — exactly where stream-k should shine.
- Pro: SYCL-TLA's `input_gemm.in` includes `--l=1 --m=3072 --k=4096 --n=3072` with a tuned config, so the comparison is against a well-optimized standard GEMM.
- Con: Not algorithm-vs-algorithm, but this is unavoidable given no stream-k exists in SYCL-TLA.

### Option B: Use only onednn as baseline (minimal)
- Just `torch.matmul` as the comparison
- Pro: Simpler (no sycl_tla import)
- Con: Misses the comparison against a tuned SYCL-TLA kernel; oneDNN may internally use its own scheduling strategies, making the comparison less transparent

### Option C: Drop comparison baseline entirely (Triton-only)
- Pro: Simplest
- Con: A stream-k benchmark with no baseline has no interpretive value — you can't tell if stream-k helps without comparing to something

**Recommendation**: Option A — use generic SYCL-TLA GEMM + onednn as dual baselines. This is the natural replacement and provides the most insight into stream-k's value proposition without requiring new C++ code.

### 1. Update `gemm_streamk_benchmark.py`
- Replace `from triton_kernels_benchmark import xetla_kernel` with `from triton_kernels_benchmark import sycl_tla_kernel`
- Change providers from `['triton', 'xetla']` to `['triton', 'onednn', 'sycl-tla']`
- Add `onednn` provider block: `torch.matmul(a, b)`
- Add `sycl-tla` provider block:
  ```python
  c = torch.zeros((M, N), device='xpu', dtype=torch.float32)
  sycl_tla_kernel.gemm(a, b, c, M, N, K, 1)
  ```
- Remove xetla provider block (no more `acc`, `cnt` tensors)
- Update styles, line_names, warmup comments

### 2. Add correctness verification
- Enable `assert_close` for both new providers

### 3. Verify shape coverage
The benchmark shape `3072 x 4096 x 3072` exists in SYCL-TLA's `input_gemm.in` (`--l=1 --m=3072 --k=4096 --n=3072`, config `PvcGemmBF16BF16FP32_RRR_1`), so it will hit a tuned config.

### 4. Consider expanding shapes
- Currently only 1 shape. Consider adding more shapes that benefit from stream-k (where `total_tiles % num_xe_cores != 0`):
  - e.g., `(1024, 4096, 1024)`, `(2048, 4096, 2048)`, `(6144, 4096, 6144)`
- Note: the Triton stream-k implementation uses `num_xe_core = torch.xpu.get_device_capability(0)['gpu_subslice_count']` so shapes that create tile counts not divisible by this will show stream-k's benefit.

### 5. Clean up main GEMM benchmark (can be deferred)
- The FIXME at `gemm_benchmark.py` line 293 keeps `(1, 3072, 3072, 4096)` in the standard GEMM benchmark "until gemm_streamk_benchmark can get better performance"
- Once xetla is removed, the special-case at lines 427-428 (routing this shape to xetla stream-k) becomes dead code — remove it

## Files to modify
- `benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py`
- `benchmarks/triton_kernels_benchmark/gemm_benchmark.py` (FIXME cleanup, can be deferred)

## Testing
- Run: `python benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py`
- Verify all providers work for the benchmark shape
- Optionally: compare Triton stream-k TFlops vs SYCL-TLA/onednn to confirm stream-k provides benefit on 3072x3072

### PR DESCRIPTION
**Pull request by slawblauciak**
SYCL-TLA doesn't have a Stream-K GEMM op available, can only compare against generic GEMM and OneDNN here

### CODE REVIEW

**Comment by anmyachev** [comment]
According to `"Stream-K is not available in SYCL-TLA. The CUTLASS framework that SYCL-TLA is based on does not implement stream-k scheduling. NVIDIA's upstream CUTLASS added stream-k later (in separate examples), but the Intel SYCL fork does not include it."` there's a chance they'll add this algorithm after upstream CUTLASS. It's okay for me to leave it as is, but we'll need to open a new issue so that when they do (if they do), we can also add this provider for comparison. @whitneywhtsang?

**Comment by whitneywhtsang** [comment]
> According to `"Stream-K is not available in SYCL-TLA. The CUTLASS framework that SYCL-TLA is based on does not implement stream-k scheduling. NVIDIA's upstream CUTLASS added stream-k later (in separate examples), but the Intel SYCL fork does not include it."` there's a chance they'll add this algorithm after upstream CUTLASS. It's okay for me to leave it as is, but we'll need to open a new issue so that when they do (if they do), we can also add this provider for comparison. @whitneywhtsang?

Agree with @anmyachev

File: benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py [hunk 1]
```diff
@@ -271,18 +271,19 @@ def benchmark(M, N, K, provider):
...
-        name = f'gemm_streamk_shape_{M}_{K}_{N}'
-        func = getattr(xetla_kernel, name)
-        xetla_fn = lambda: func(a, b, c, acc, cnt)
-        torch_fn = lambda: torch.matmul(a, b).to(torch.float32)
+    elif provider == 'sycl-tla':
+        # NOTE: SYCL-TLA doesn't have a stream-k matmul op
+        func = getattr(sycl_tla_kernel, 'gemm')
+
+        def sycl_tla_invoker():
+            c = torch.zeros((M, N), device='xpu', dtype=torch.float32)
```
Comment (anmyachev) [inline benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py]: Let's move out this allocation from the benchmark function.

Comment (slawblauciak) [reply benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py]: Ok. But do note that the same pattern already exists in gemm_benchmark

**Review by whitneywhtsang — APPROVED** [review]

File: benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py (hunk 1 above)
Comment (anmyachev) [reply benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py]: gemm_benchmark appears to have an unfair comparison. @slawblauciak could you fix it as well?

**Comment by anmyachev** [comment]
> According to `"Stream-K is not available in SYCL-TLA. The CUTLASS framework that SYCL-TLA is based on does not implement stream-k scheduling. NVIDIA's upstream CUTLASS added stream-k later (in separate examples), but the Intel SYCL fork does not include it."` there's a chance they'll add this algorithm after upstream CUTLASS. It's okay for me to leave it as is, but we'll need to open a new issue so that when they do (if they do), we can also add this provider for comparison. @whitneywhtsang?

@slawblauciak have you opened a new issue?

File: benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py (hunk 1 above)
Comment (slawblauciak) [reply benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py]: Meaning? Unfair, as in because of the alloc? I can fix it (in a separate PR)

Comment (anmyachev) [reply benchmarks/triton_kernels_benchmark/gemm_streamk_benchmark.py]: > Unfair, as in because of the alloc?

yes

> I can fix it (in a separate PR)

thanks!

**Review by anmyachev — APPROVED** [review]

**Comment by slawblauciak** [comment]
> > According to `"Stream-K is not available in SYCL-TLA. The CUTLASS framework that SYCL-TLA is based on does not implement stream-k scheduling. NVIDIA's upstream CUTLASS added stream-k later (in separate examples), but the Intel SYCL fork does not include it."` there's a chance they'll add this algorithm after upstream CUTLASS. It's okay for me to leave it as is, but we'll need to open a new issue so that when they do (if they do), we can also add this provider for comparison. @whitneywhtsang?
> 
> @slawblauciak have you opened a new issue?

No, I will create one.
